# Рынок внутренней интернет-торговли в России

- [**Отчет «Рынок внутренней интернет-торговли в России»**](<https://drive.google.com/drive/folders/1VK1Br5ix295rM_5od4lJjhicuZfGc8Ae?usp=sharing>)


- [**Презентация**](<https://drive.google.com/file/d/11DWSaxajvNCbMpaaIZuUoZ3Wnea0vglw/view?usp=sharing>)

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-и-обзор-данных" data-toc-modified-id="Загрузка-и-обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка и обзор данных</a></span><ul class="toc-item"><li><span><a href="#Подключение-библиотек" data-toc-modified-id="Подключение-библиотек-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Подключение библиотек</a></span></li><li><span><a href="#Подключение-к-базе-данных" data-toc-modified-id="Подключение-к-базе-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Подключение к базе данных</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Задачи-исследования" data-toc-modified-id="Задачи-исследования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Задачи исследования</a></span><ul class="toc-item"><li><span><a href="#Динамика-оборота-и-среднего-чека,-сравнение-в-аналогичных-периодах" data-toc-modified-id="Динамика-оборота-и-среднего-чека,-сравнение-в-аналогичных-периодах-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Динамика оборота и среднего чека, сравнение в аналогичных периодах</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Региональное-распределение-рынка-по-обороту" data-toc-modified-id="Региональное-распределение-рынка-по-обороту-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Региональное распределение рынка по обороту</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Определение-ТОП-5-категорий-товаров-по-обороту" data-toc-modified-id="Определение-ТОП-5-категорий-товаров-по-обороту-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Определение ТОП-5 категорий товаров по обороту</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.3.1"><span class="toc-item-num">3.3.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Сравнение-среднего-чека-выбранного-региона-со-средним-чеком-группы-регионов-по-численности-населения" data-toc-modified-id="Сравнение-среднего-чека-выбранного-региона-со-средним-чеком-группы-регионов-по-численности-населения-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Сравнение среднего чека выбранного региона со средним чеком группы регионов по численности населения</a></span><ul class="toc-item"><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.4.1"><span class="toc-item-num">3.4.1&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li><li><span><a href="#Создание-представления" data-toc-modified-id="Создание-представления-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Создание представления</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

# Рынок внутренней интернет-торговли в России

**Описание проекта**:

Создание сравнительного отчета для аналитики рынка внутренней интернет-торговли в России 2022 VS 2021.

**Цель исследования**:
- Обзор рынка внутренней интернет-торговли в России в 2021 - 2022 г.
- Динамика оборота и среднего чека, сравнение в аналогичных периодах.
- Региональное распределение рынка по обороту.
- Определение ТОП-5 категорий товаров по обороту.
- Определение доли регионов и категорий товаров в обороте, сравнение доли в эквивалентных периодах 2021 и 2022 г. 
- Сравнение среднего чека выбранного региона со средним чеком группы регионов по численности населения.

**Ход исследования**:

Из открытых источников собраны данные об обороте и среднем чеке внутренней интернет-торговли по категориям товаров в регионах за 2021 - 2022, а также о численности населения регионов РФ на 01.01.2022:
- обороты интернет-торговли по категориям товаров в регионах (https://akit.ru/analytics/trade);
- данные по численности населения регионов РФ на 01.01.2022 (https://www.statdata.ru/largest_regions_russia).


Сначала необходимо подготовить датасет и создать базу данных в MySQL. На этапе предобработки нужно проверить данные на наличие отсутствующих значений. Далее написать необходимые SQL-запросы и создать представление для решения задач исследования. Затем на основании представления разработать отчет в Power BI Desktop. 

Таким образом, исследование пройдет в шесть этапов:

- Загрузка и обзор данных;
- Предобработка данных;
- Задачи исследования;
- Разработка отчета в Power BI Desktop;
- Общий вывод
- [Отчет](<https://drive.google.com/drive/folders/1VK1Br5ix295rM_5od4lJjhicuZfGc8Ae?usp=sharing>)
- [Презентация](<https://drive.google.com/file/d/11DWSaxajvNCbMpaaIZuUoZ3Wnea0vglw/view?usp=sharing>)

**Детали статистики:**
- В исследование рынка торговли включены только покупки физических товаров, осуществлённые в интернет-магазинах.
- Приобретение контента, авиабилетов и другие электронные услуги в исследовании не учитывались.
- В общем объёме рынка включена стоимость услуг по доставке товаров в случаях, когда данная услуга являлась неотъемлемой частью договоров дистанционной купли-продажи товаров.
- Покупки в российских интернет-магазинах осуществляются путём оплаты (пост- или предоплаты) посредством как безналичных, так и наличных расчётов.

## Загрузка и обзор данных

### Подключение библиотек

In [1]:
import pandas as pd

import pymysql
from sqlalchemy import create_engine

### Подключение к базе данных

Создадим подключение к базе данных `market`. База данных создана в MySQL в клиенте DBeaver.

In [2]:
cnx = create_engine('mysql+pymysql://root:aFNXgCdmVmXfc@localhost/market')  

**Таблица `trade`**

- Содержит данные об обороте и среднем чеке по датам, идентификаторе региона и идентификаторе категории товаров внутренней интернет-торговли РФ.
- Проведена нормализация данных из источника данных об обороте и среднем чеке. 
- Первичным ключом является идентификатор (поле `id`) записи в таблице.

Создадим таблицу `trade`.

In [3]:
query = '''
CREATE TABLE trade (
    id INT PRIMARY KEY AUTO_INCREMENT,
    region VARCHAR(40),
    type_trade VARCHAR(30),
    date_trade DATE,
    category VARCHAR(40),
    turnover DECIMAL(8, 2),
    avg_check INT
);
'''

Выполнение запроса.

In [126]:
with cnx.connect() as connection:
    connection.execute(query)

**Таблица `region_data_book`** 

- Таблица-справочник регионов. В исходных данных из нескольких источников регионы называются по-разному. Данная таблица содержит наименование региона РФ согласно источнику данных об обороте и среднем чеке (поле `region_trade`), наименование региона согласно источнику данных о численности населения (поле `region_population`) и идентификатор региона (поле `id`), который совпадает с идентификатором региона карты, используемой для создания отчета в Power BI;
- Первичным ключом является поле идентификатор региона `id`.

Создадим таблицу `region_data_book`.

In [5]:
query = '''
CREATE TABLE region_data_book (
    id INT PRIMARY KEY,
    region_trade VARCHAR(40),
    region_population VARCHAR(40)
);
'''

Выполнение запроса.

In [ ]:
with cnx.connect() as connection:
    connection.execute(query)

**Таблица `population_base`** 

- Содержит данные о численности населения регионов РФ.
- Первичным ключом является поле регион (`region`).

Создадим таблицу `population_base`.

In [6]:
query = '''
CREATE TABLE population_base (
    region VARCHAR(40) PRIMARY KEY,
    population INT
);
'''

Выполнение запроса.

In [ ]:
with cnx.connect() as connection:
    connection.execute(query)

**Таблица `category_base`**

- Содержит данные о категориях товаров.
- Первичным ключом является поле идентификатор категории товаров (`category_id`).

Создадим таблицу `category_base`.

In [7]:
query = '''
CREATE TABLE category_base (
    category_id INT PRIMARY KEY,
    category VARCHAR(40)
);
'''

Выполнение запроса.

In [ ]:
with cnx.connect() as connection:
    connection.execute(query)

Добавим внешние ключи в таблицу `trade`.
- поле регион `region_id` (ссылается на поле `id` таблицы `region_data_book`);

In [8]:
query = '''
ALTER TABLE trade
ADD FOREIGN KEY (region_id) REFERENCES region_data_book(id);
'''

Выполнение запроса.

In [ ]:
with cnx.connect() as connection:
    connection.execute(query)

- поле `category_id` (ссылается на поле `category_id` таблицы `category_base`).

In [ ]:
query = '''
ALTER TABLE trade
ADD FOREIGN KEY (category_id) REFERENCES category_base(category_id);
'''

Выполнение запроса.

In [ ]:
with cnx.connect() as connection:
    connection.execute(query)

Добавим внешний ключ в таблицу `region_data_book`:
- поле `region_population` (ссылается на поле `region` таблицы `population_base`).

In [ ]:
query = '''
ALTER TABLE region_data_book
ADD FOREIGN KEY (region_population) REFERENCES population_base (region);
'''

Выполнение запроса.

In [ ]:
with cnx.connect() as connection:
    connection.execute(query)

* Отношения таблиц в базе данных отражены на [**ER-диаграмме**](<https://drive.google.com/file/d/1tqVNVSlfmJiTRU_kAVEdYnr1-wq71hIV/view?usp=sharing>).
* Данные загружены в таблицы базы в клиенте DBeaver.

* Создадим функцию `getQuery()` для вывода данных SQL-запроса в датафрейм.

In [3]:
def getQuery(query):
    df = pd.read_sql(query, cnx)
    return df

* Напишем запрос на выгрузку первых 5 строк таблицы `trade`.

In [60]:
query = '''
SELECT * 
  FROM trade
 LIMIT 5
'''

Применим функцию `getQuery()`.

In [61]:
getQuery(query)

id  region_id  category_id  date_trade  turnover  avg_check
0   1          8            1  2021-01-01    255.86       2162
1   2          9            1  2021-01-01    277.04       2751
2   3         83            1  2021-01-01    354.88       2858
3   4         86            1  2021-01-01    608.76       1702
4   5         22            1  2021-01-01    209.59       2261

* Посчитаем количество строк в таблице `trade`.

In [5]:
query = '''
SELECT COUNT(*) AS number_of_rows  -- возвращает число записей в таблице
  FROM trade
'''
getQuery(query)

number_of_rows
0           11560

* Cтруктура таблицы `trade`:
    * `id` - идентификатор записи в таблице;
    * `region_id` - идентификатор региона;
    * `category_id` - идентификатор категории товаров;
    * `date_trade` - дата начала квартала;
    * `turnover` - оборот, млн руб.;
    * `avg_check` - средний чек, руб.

* Тип данных:
    * в полях `id`, `region_id`, `category_id`, `avg_check`  - `int`;
    * в поле `date_trade` — `date`;
    * в поле `turnover` — `decimal`.

* Напишем запрос на выгрузку первых 5 строк таблицы `region_data_book`.

In [6]:
query = '''
SELECT * 
  FROM region_data_book
 LIMIT 5 -- вывод первых 5 строк
'''
getQuery(query)

id           region_trade          region_population
0   1      Республика Адыгея          Республика Адыгея
1   3         Алтайский край             Алтайский край
2   4       Амурская область           Амурская область
3   5  Архангельская область  Архангельская обл.без НАО
4   6   Астраханская область       Астраханская область

* Посчитаем количество строк в таблице `region_data_book`.

In [7]:
query = '''
SELECT COUNT(*) AS number_of_rows  -- возвращает число записей в таблице
  FROM region_data_book
'''
getQuery(query)

number_of_rows
0              85

* Cтруктура таблицы `region_data_book`:
    * `id` — идентификатор региона;
    * `region_trade` — наименование региона РФ согласно источнику данных об обороте и среднем чеке;
    * `region_population` — наименование региона согласно таблице источнику данных о численности населения.
* Тип данных:
    * в поле `id` — `int`;
    * в полях `region_trade`, `region_population` — `varchar`.

* Напишем запрос на выгрузку первых 5 строк таблицы `population_base`.

In [12]:
query = '''
SELECT * 
  FROM population_base
 LIMIT 5 -- вывод первых 5 строк
'''

In [42]:
getQuery(query)

region  population
0             Алтайский край     2268179
1           Амурская область      772525
2  Архангельская обл.без НАО     1069782
3       Астраханская область      989430
4       Белгородская область     1531917

* Посчитаем количество строк в таблице `population_base`.

In [13]:
query = '''
SELECT COUNT(*) AS number_of_rows  -- возвращает число записей в таблице
  FROM population_base
'''
getQuery(query)

number_of_rows
0              85

* Cтруктура таблицы `population_base`:
    * `region` — название региона;
    * `population` — численность населения региона РФ на 01.01.2022, чел.
* Тип данных:
    * в поле `region` - `varchar`;
    * в поле `population` — `int`.

* Напишем запрос на выгрузку таблицы `category_base`.

In [14]:
query = '''
SELECT * 
  FROM category_base
'''
getQuery(query)

category_id                       category
0             1  Автозапчасти и автоаксессуары
1             2                     Аксессуары
2             3                         Аптеки
3             4                    Зоомагазины
4             5  Инструменты и садовая техника
5             6                          Книги
6             7             Красота и здоровье
7             8       Мебель и товары для дома
8             9                 Одежда и обувь
9            10                Подарки и цветы
10           11               Продукты питания
11           12                  Прочие товары
12           13                    Спорттовары
13           14               Товары для детей
14           15          Товары для творчества
15           16     Цифровая и бытовая техника
16           17       Ювелирные изделия и часы

* Посчитаем количество строк в таблице `category_base`.

In [15]:
query = '''
SELECT COUNT(*) AS number_of_rows  -- возвращает число записей в таблице
  FROM category_base
'''
getQuery(query)

number_of_rows
0              17

* Cтруктура таблицы `category_base`:
    * `category_id` — идентификатор категории товаров;
    * `category` — наименование категории товаров.
* Тип данных:
    * в поле `category_id` — `int`;
    * в поле `category` — `varchar`.

### Вывод

Данные содержат информацию об обороте и среднем чеке внутренней интернет-торговли по 17-ти категориям товаров в регионах, а также о численности населения регионов РФ.

## Предобработка данных

- При подготовке датасета для таблицы `trade` в исходных данных была обнаружена ошибка: завышенное значение среднего чека для категории "Аксессуары" в Московской области за май 2021 г. - 196942 руб. Для сравнения, в апреле - 1538 руб., в июне - 1911 руб. Приняли, что в мае было 1969,42 руб., соответственно, средний чек за второй квартал 2021 г. составил 1806 руб.
- Посмотрим на значение среднего чека за второй квартал 2021 г. для категории "Аксессуары" в Московской области в базе данных.

In [34]:
query = '''
WITH trade_details AS (
    SELECT *
      FROM trade
     WHERE YEAR(date_trade) = 2021    -- 2021 год
        AND QUARTER(date_trade) = 2   -- 2й квартал
),
region_details AS (
    SELECT
        id,
        region_trade
      FROM region_data_book
     WHERE region_trade = 'Московская область'
),
category_details AS (
    SELECT
        category_id,
        category
      FROM category_base
     WHERE category = 'Аксессуары'
)
SELECT date_trade,
       avg_check,
       region_trade,
       category
 FROM trade_details
      INNER JOIN region_details
      ON trade_details.region_id = region_details.id
      
      INNER JOIN category_details
      ON trade_details.category_id = category_details.category_id;
'''
getQuery(query)

date_trade  avg_check        region_trade    category
0  2021-04-01       1806  Московская область  Аксессуары

Значение среднего чека в базе данных корректное.

* Проверим наличие отсутствующих значений в полях `turnover`, `avg_check` в таблице `trade` с помощью запроса.

In [19]:
query = '''
SELECT *
  FROM trade 
 WHERE turnover IS NULL
   OR avg_check IS NULL;
'''
getQuery(query)

Empty DataFrame
Columns: [id, region_id, category_id, date_trade, turnover, avg_check]
Index: []

Нет отсутствующих значений в этих полях.

* Проверим наличие отсутствующих значений в поле `population` в таблице `population_base` с помощью запроса.

In [20]:
query = '''
SELECT *
  FROM population_base 
 WHERE population IS NULL;
'''
getQuery(query)

Empty DataFrame
Columns: [region, population, population_group]
Index: []

Нет отсутствующих значений в этом поле.

* Проверим, за какой период есть данные в таблице `trade` с помощью запроса.

In [22]:
query = '''
SELECT 
    DISTINCT QUARTER(date_trade) AS quarter_trade,  -- квартал
    YEAR(date_trade) AS year_trade -- год
  FROM trade;
'''
getQuery(query)

quarter_trade  year_trade
0              1        2021
1              2        2021
2              3        2021
3              4        2021
4              1        2022
5              2        2022
6              3        2022
7              4        2022

Данные за 2021 и 2022 год.

### Вывод

Данные за 2021 и 2022 год. Нет отсутствующих значений в числовых полях таблиц.

## Задачи исследования

### Динамика оборота и среднего чека, сравнение в аналогичных периодах

* Найдем оборот (`total_turnover`), средний чек (`avg_check`), прирост оборота год к году (`yoy_turnover`), прирост среднего чека год к году (`yoy_avg_check`) в 2022 и 2021 годах.

In [29]:
query = '''
WITH trade_details AS (
        SELECT YEAR(date_trade) AS year_trade,
               ROUND(SUM(turnover), 0) AS total_turnover,
               ROUND(AVG(avg_check), 0) AS avg_check
          FROM trade
         GROUP BY YEAR(date_trade)
)
SELECT year_trade,
       total_turnover,       
       -- считаем прирост оборота с помощью оконной функции смещения
       ROUND((total_turnover - LAG(total_turnover, 1) OVER (ORDER BY year_trade)) /
             LAG(total_turnover, 1) OVER (ORDER BY year_trade)*100, 0
       ) AS yoy_turnover,
       avg_check,       
       -- считаем прирост среднего чека с помощью оконной функции смещения
       ROUND((avg_check - LAG(avg_check, 1) OVER (ORDER BY year_trade)) /
             LAG(avg_check, 1) OVER (ORDER BY year_trade)*100, 0
       ) AS yoy_avg_check
  FROM trade_details
 ORDER BY year_trade DESC;
'''
getQuery(query).fillna('-')

year_trade  total_turnover yoy_turnover  avg_check yoy_avg_check
0        2022       4809063.0         33.0     3447.0          20.0
1        2021       3604300.0            -     2868.0             -

- Оборот внутренней интернет-торговли в России за 2022 год увеличился на 33% по сравнению с показателем 2021 года и составил 4,81 трлн рублей.
- Средний чек увеличился на 20% по сравнению с 2021 годом и составил 3 447 руб.

* Найдем оборот (`total_turnover`), средний чек (`avg_check`), прирост оборота год к году (`yoy_turnover`), прирост среднего чека год к году (`yoy_avg_check`) по аналогичным кварталам в 2022 и 2021 годах.

In [31]:
query = '''
WITH trade_details AS (
    SELECT QUARTER(date_trade) AS quarter_trade,
           YEAR(date_trade) AS year_trade,
           ROUND(SUM(turnover), 0) AS total_turnover,
           ROUND(AVG(avg_check), 0) AS avg_check
      FROM trade
     GROUP BY QUARTER(date_trade),
              YEAR(date_trade)
)
SELECT year_trade,
       quarter_trade,
       total_turnover,
       -- считаем прирост оборота с помощью оконной функции смещения
       ROUND((total_turnover - LAG(total_turnover, 4) OVER (ORDER BY year_trade, quarter_trade)) /
             LAG(total_turnover, 4) OVER (ORDER BY year_trade, quarter_trade)*100, 0
       ) AS yoy_turnover,
       avg_check,
       -- считаем прирост среднего чека с помощью оконной функции смещения
       ROUND((avg_check - LAG(avg_check, 4) OVER (ORDER BY year_trade, quarter_trade)) /
             LAG(avg_check, 4) OVER (ORDER BY year_trade, quarter_trade)*100, 0
       ) AS yoy_avg_check
  FROM trade_details
 ORDER BY year_trade DESC,
          quarter_trade ASC;
'''
getQuery(query).fillna('-')

year_trade  quarter_trade  total_turnover yoy_turnover  avg_check  \
0        2022              1       1114771.0         63.0     3233.0   
1        2022              2       1065794.0         38.0     3336.0   
2        2022              3       1191699.0         32.0     3579.0   
3        2022              4       1436799.0         15.0     3639.0   
4        2021              1        684178.0            -     2662.0   
5        2021              2        771814.0            -     2819.0   
6        2021              3        903380.0            -     2942.0   
7        2021              4       1244929.0            -     3048.0   

  yoy_avg_check  
0          21.0  
1          18.0  
2          22.0  
3          19.0  
4             -  
5             -  
6             -  
7             -

- В динамике оборота наблюдаем сезонность по кварталам. Наибольший оборот в 4-м квартале. В 1-м квартале 2022 г. оборот был выше оборота в 1-м квартале 2021 г. на 63%, в 4-м квартале 2022 г. оборот был выше оборота в 4-м квартале 2021 г. на 15%.<br>

- В 1-м квартале 2022 г. средний чек был выше среднего чека в 1-м квартале 2021 г. на 21%, в 4-м квартале 2022 г. средний чек был выше среднего чека в 4-м квартале 2021 г. на 19%.

* Найдем оборот (`total_turnover`), средний чек (`avg_check`), прирост оборота квартал к кварталу (`qoq_turnover`), прирост среднего чека квартал к кварталу (`qoq_avg_check`) в 2022 и 2021 годах.

In [4]:
query = '''
WITH trade_details AS (
    SELECT QUARTER(date_trade) AS quarter_trade,
           YEAR(date_trade) AS year_trade,
           ROUND(SUM(turnover), 0) AS total_turnover,
           ROUND(AVG(avg_check), 1) AS avg_check
      FROM trade
     GROUP BY QUARTER(date_trade),
              YEAR(date_trade)
)
SELECT year_trade,
       quarter_trade,
       total_turnover,
       -- считаем прирост оборота с помощью оконной функции смещения
       ROUND((total_turnover - LAG(total_turnover, 1) OVER (ORDER BY year_trade, quarter_trade)) /
             LAG(total_turnover, 1) OVER (ORDER BY year_trade, quarter_trade)*100, 0
       ) AS qoq_turnover,
       avg_check,
       -- считаем прирост среднего чека с помощью оконной функции смещения
       ROUND((avg_check - LAG(avg_check, 1) OVER (ORDER BY year_trade, quarter_trade)) /
             LAG(avg_check, 1) OVER (ORDER BY year_trade, quarter_trade)*100, 0
       ) AS qoq_avg_check
  FROM trade_details
 ORDER BY year_trade DESC,
          quarter_trade ASC;
'''
getQuery(query).fillna('-')

year_trade  quarter_trade  total_turnover qoq_turnover  avg_check  \
0        2022              1       1114771.0        -10.0     3233.2   
1        2022              2       1065794.0         -4.0     3336.0   
2        2022              3       1191699.0         12.0     3579.2   
3        2022              4       1436799.0         21.0     3639.4   
4        2021              1        684178.0            -     2662.1   
5        2021              2        771814.0         13.0     2818.6   
6        2021              3        903380.0         17.0     2941.9   
7        2021              4       1244929.0         38.0     3047.8   

  qoq_avg_check  
0           6.0  
1           3.0  
2           7.0  
3           2.0  
4             -  
5           6.0  
6           4.0  
7           4.0

- В 2021 г. оборот увеличивался с 1-го по 4-й квартал. В 2022 г. в 1-м и 2-м кварталах наблюдаем  падение оборота и рост с 3-го квартала. Самый большой прирост оборота в 4-м квартале (в 2021 и в 2022).

- Средний чек увеличивался с каждым кварталом 2021 и 2022 г. в среднем на 4,5%. 

#### Вывод

- Оборот внутренней интернет-торговли в России за 2022 год увеличился на 33% по сравнению с показателем 2021 года и составил 4,81 трлн рублей.
- Средний чек увеличился на 20% по сравнению с 2021 годом и составил 3 447 руб.


- На потребительское поведение в 2022 году повлияли политические события.
  - Снижение оборота в 1-м и 2-м квартале может быть обусловлено тем, что: 
    - Сократились объемы производства, строительства, недвижимости.
    - Многие компании покинули российский рынок.
    - Кто-то в начале весны стал покупать меньше из-за неопределенности и возможности потерять работу, сложности строить планы на будущее. 
  - Летом покупательская активность в интернете постепенно начала восстанавливаться.
  - В 4м-квартале покупательская активность увеличилась - влияние сезонного повышения спроса в связи с распродажами, покупкой подарков к новогодним праздникам.

### Региональное распределение рынка по обороту

* Найдем оборот (`total_turnover`), долю в обороте (`turnover_share`), изменение доли в обороте (`yoy_turn_share`), средний чек (`avg_check`), прирост оборота год к году (`yoy_turnover`), прирост среднего чека год к году (`yoy_avg_check`) в 2022 и 2021 годах и вывод топ-10 регионов по обороту.

**2021:**

In [46]:
query = '''
WITH trade_region AS (
    SELECT YEAR(date_trade) AS year_trade,
           region_id,
           ROUND(SUM(turnover), 0) AS total_turnover,
           ROUND(AVG(avg_check), 0) AS avg_check
      FROM trade
     GROUP BY region_id, YEAR(date_trade)
),
region_details AS (
    SELECT id,
           region_trade
       FROM region_data_book
),
turnover_details AS (
    SELECT YEAR(date_trade) AS year_trade,
           ROUND(SUM(turnover), 0) AS year_total_turnover -- считаем годовой оборот
      FROM trade
     GROUP BY YEAR(date_trade)
),
trade_region_details AS (
SELECT trade_region.year_trade,
       region_trade,
       total_turnover,
       -- считаем долю в обороте
       ROUND(total_turnover*100 / year_total_turnover, 1) AS turnover_share,       
       -- считаем изменение доли в обороте с помощью оконной функции смещения
       ROUND((total_turnover/year_total_turnover - LAG(total_turnover/year_total_turnover, 1) OVER (
                                                       PARTITION BY region_trade  -- определение партиций по региону
                                                       ORDER BY year_trade))*100, 1
       ) AS yoy_turn_share,      
       -- считаем прирост оборота с помощью оконной функции смещения
       ROUND((total_turnover - LAG(total_turnover, 1) OVER (
                                   PARTITION BY region_trade -- определение партиций по региону
                                   ORDER BY year_trade)) / LAG(total_turnover, 1) OVER (
                                                               PARTITION BY region_trade -- определение партиций по региону
                                                               ORDER BY year_trade)*100, 0
       ) AS yoy_turnover,             
       avg_check,        
       -- считаем прирост среднего чека с помощью оконной функции смещения
       ROUND((avg_check - LAG(avg_check, 1) OVER (
                              PARTITION BY region_trade  -- определение партиций по региону
                              ORDER BY year_trade)) / LAG(avg_check, 1) OVER (
                                                          PARTITION BY region_trade  -- определение партиций по региону
                                                          ORDER BY year_trade)*100, 0
       ) AS yoy_avg_check     
  FROM trade_region
       INNER JOIN region_details AS r1
       ON trade_region.region_id = r1.id
       
       LEFT JOIN turnover_details AS t1
       ON trade_region.year_trade = t1.year_trade
),
trade_region_rnk AS (
SELECT *,
       -- считаем ранг с помощью ранжирующей оконной функции 
       RANK() OVER(ORDER BY total_turnover DESC) AS rnk 
  FROM trade_region_details
 WHERE trade_region_details.year_trade = 2021
)
SELECT *
  FROM trade_region_rnk
 WHERE rnk <= 10; -- вывод ТОП-10 регионов по обороту
'''
getQuery(query).fillna('-')

year_trade             region_trade  total_turnover  turnover_share  \
0        2021                   Москва        749542.0            20.8   
1        2021       Московская область        388609.0            10.8   
2        2021          Санкт-Петербург        266847.0             7.4   
3        2021       Краснодарский край        157871.0             4.4   
4        2021     Свердловская область         95149.0             2.6   
5        2021       Ростовская область         87181.0             2.4   
6        2021    Нижегородская область         72525.0             2.0   
7        2021        Самарская область         70532.0             2.0   
8        2021     Республика Татарстан         68267.0             1.9   
9        2021  Республика Башкортостан         57635.0             1.6   

  yoy_turn_share yoy_turnover  avg_check yoy_avg_check  rnk  
0              -            -     4333.0             -    1  
1              -            -     3457.0             -    2  
2              -            -     3000.0             -    3  
3              -            -     3537.0             -    4  
4              -            -     2826.0             -    5  
5              -            -     2873.0             -    6  
6              -            -     2638.0             -    7  
7              -            -     2967.0             -    8  
8              -            -     2960.0             -    9  
9              -            -     2665.0             -   10

**ТОП-10 регионов по доле в обороте в 2021:**
  
  1. Москва - доля 20,8%;
  2. Московская область - доля 10,8%;
  3. Санкт-Петербург - доля 7,4%;
  4. Краснодарский край - доля 4,4%;
  5. Свердловская область - доля 2,6%;
  6. Ростовская область - доля 2,4%;
  7. Нижегородская область - доля 2,0%;
  8. Самарская область - доля 2,0%;
  9. Республика Татарстан - доля 1,9%;
  10. Республика Башкортостан - доля 1,6%.

**2022:**

In [47]:
query = '''
WITH trade_region AS (
    SELECT YEAR(date_trade) AS year_trade,
           region_id,
           ROUND(SUM(turnover), 0) AS total_turnover,
           ROUND(AVG(avg_check), 0) AS avg_check
      FROM trade
     GROUP BY region_id, YEAR(date_trade)
),
region_details AS (
    SELECT id,
           region_trade
       FROM region_data_book
),
turnover_details AS (
    SELECT YEAR(date_trade) AS year_trade,
           ROUND(SUM(turnover), 0) AS year_total_turnover -- считаем годовой оборот
      FROM trade
     GROUP BY YEAR(date_trade)
),
trade_region_details AS (
SELECT trade_region.year_trade,
       region_trade,
       total_turnover,
       -- считаем долю в обороте
       ROUND(total_turnover*100 / year_total_turnover, 1) AS turnover_share,       
       -- считаем изменение доли в обороте с помощью оконной функции смещения
       ROUND((total_turnover/year_total_turnover - LAG(total_turnover/year_total_turnover, 1) OVER (
                                                       PARTITION BY region_trade  -- определение партиций по региону
                                                       ORDER BY year_trade))*100, 1
       ) AS yoy_turn_share,      
       -- считаем прирост оборота с помощью оконной функции смещения
       ROUND((total_turnover - LAG(total_turnover, 1) OVER (
                                   PARTITION BY region_trade -- определение партиций по региону
                                   ORDER BY year_trade)) / LAG(total_turnover, 1) OVER (
                                                               PARTITION BY region_trade -- определение партиций по региону
                                                               ORDER BY year_trade)*100, 0
       ) AS yoy_turnover,             
       avg_check,        
       -- считаем прирост среднего чека с помощью оконной функции смещения
       ROUND((avg_check - LAG(avg_check, 1) OVER (
                              PARTITION BY region_trade  -- определение партиций по региону
                              ORDER BY year_trade)) / LAG(avg_check, 1) OVER (
                                                          PARTITION BY region_trade  -- определение партиций по региону
                                                          ORDER BY year_trade)*100, 0
       ) AS yoy_avg_check     
  FROM trade_region
       INNER JOIN region_details AS r1
       ON trade_region.region_id = r1.id
       
       LEFT JOIN turnover_details AS t1
       ON trade_region.year_trade = t1.year_trade
),
trade_region_rnk AS (
SELECT *,
       -- считаем ранг с помощью ранжирующей оконной функции 
       RANK() OVER(ORDER BY total_turnover DESC) AS rnk 
  FROM trade_region_details
 WHERE trade_region_details.year_trade = 2022
)
SELECT *
  FROM trade_region_rnk
 WHERE rnk <= 10; -- вывод ТОП-10 регионов по обороту
'''
getQuery(query)

year_trade           region_trade  total_turnover  turnover_share  \
0        2022                 Москва        940956.0            19.6   
1        2022     Московская область        518647.0            10.8   
2        2022        Санкт-Петербург        338847.0             7.0   
3        2022     Краснодарский край        226413.0             4.7   
4        2022     Ростовская область        125469.0             2.6   
5        2022   Свердловская область        124909.0             2.6   
6        2022  Нижегородская область        101365.0             2.1   
7        2022      Самарская область         95304.0             2.0   
8        2022    Воронежская область         94726.0             2.0   
9        2022   Республика Татарстан         87745.0             1.8   

   yoy_turn_share  yoy_turnover  avg_check  yoy_avg_check  rnk  
0            -1.2          26.0     4473.0            3.0    1  
1             0.0          33.0     3906.0           13.0    2  
2            -0.4          27.0     3719.0           24.0    3  
3             0.3          43.0     3980.0           13.0    4  
4             0.2          44.0     3697.0           29.0    5  
5             0.0          31.0     3420.0           21.0    6  
6             0.1          40.0     3303.0           25.0    7  
7             0.0          35.0     3480.0           17.0    8  
8             0.5          80.0     3341.0           26.0    9  
9            -0.1          29.0     3341.0           13.0   10

**ТОП-10 регионов по доле в обороте в 2022:**
  
  1. Москва - доля 19,6% (уменьшилась на 1,2 п.п. за год);
  2. Московская область - доля 10,8% (не изменилась за год);
  3. Санкт-Петербург - доля 7% (уменьшилась на 0,4 п.п. за год);
  4. Краснодарский край - доля 4,7% (увеличилась на 0,3 п.п. за год);
  5. Ростовская область - доля 2,6% (увеличилась на 0,2 п.п. за год);
  6. Свердловская область - доля 2,6% (не изменилась за год);
  7. Нижегородская область - доля 2,1% (увеличилась на 0,1 п.п. за год);
  8. Самарская область - доля 2% (не изменилась за год);
  9. **Воронежская область** - доля 2% (увеличилась на **0,5 п.п.** за год - **самый высокий прирост доли**);
  10. Республика Татарстан - доля 1,8% (уменьшилась на 0,1 п.п. за год).

- В 2021 г. в ТОП-10 входила республика Башкортостан, а Воронежской области не было в ТОП-10. В 2022 г. Воронежская область показала самый высокий прирост доли.
- Ростовская область в 2022 г. догнала Свердловскую область.
- Республика Татарстан опустилась на 10-е место.

* Посмотрим, какое место заняла Воронежская область в 2021 г.

In [48]:
query = '''
WITH trade_region AS (
    SELECT YEAR(date_trade) AS year_trade,
           region_id,
           ROUND(SUM(turnover), 0) AS total_turnover,
           ROUND(AVG(avg_check), 0) AS avg_check
      FROM trade
     GROUP BY region_id, YEAR(date_trade)
),
region_details AS (
    SELECT id,
           region_trade
       FROM region_data_book
),
turnover_details AS (
    SELECT YEAR(date_trade) AS year_trade,
           ROUND(SUM(turnover), 0) AS year_total_turnover -- считаем годовой оборот
      FROM trade
     GROUP BY YEAR(date_trade)
),
trade_region_details AS (
SELECT trade_region.year_trade,
       region_trade,
       total_turnover,
       -- считаем долю в обороте
       ROUND(total_turnover*100 / year_total_turnover, 1) AS turnover_share,       
       -- считаем изменение доли в обороте с помощью оконной функции смещения
       ROUND((total_turnover/year_total_turnover - LAG(total_turnover/year_total_turnover, 1) OVER (
                                                       PARTITION BY region_trade  -- определение партиций по региону
                                                       ORDER BY year_trade))*100, 1
       ) AS yoy_turn_share,      
       -- считаем прирост оборота с помощью оконной функции смещения
       ROUND((total_turnover - LAG(total_turnover, 1) OVER (
                                   PARTITION BY region_trade -- определение партиций по региону
                                   ORDER BY year_trade)) / LAG(total_turnover, 1) OVER (
                                                               PARTITION BY region_trade -- определение партиций по региону
                                                               ORDER BY year_trade)*100, 0
       ) AS yoy_turnover,             
       avg_check,        
       -- считаем прирост среднего чека с помощью оконной функции смещения
       ROUND((avg_check - LAG(avg_check, 1) OVER (
                              PARTITION BY region_trade  -- определение партиций по региону
                              ORDER BY year_trade)) / LAG(avg_check, 1) OVER (
                                                          PARTITION BY region_trade  -- определение партиций по региону
                                                          ORDER BY year_trade)*100, 0
       ) AS yoy_avg_check     
  FROM trade_region
       INNER JOIN region_details AS r1
       ON trade_region.region_id = r1.id
       
       LEFT JOIN turnover_details AS t1
       ON trade_region.year_trade = t1.year_trade
),
trade_region_rnk AS (
SELECT *,
       -- считаем ранг с помощью ранжирующей оконной функции 
       RANK() OVER(ORDER BY total_turnover DESC) AS rnk 
  FROM trade_region_details
 WHERE trade_region_details.year_trade = 2021
)
SELECT *
  FROM trade_region_rnk
 WHERE region_trade = 'Воронежская область'
'''
getQuery(query).fillna('-')

year_trade         region_trade  total_turnover  turnover_share  \
0        2021  Воронежская область         52569.0             1.5   

  yoy_turn_share yoy_turnover  avg_check yoy_avg_check  rnk  
0              -            -     2646.0             -   14

В 2021 г. Воронежская область была на 14-м месте.

* Найдем 5 регионов с наибольшими темпами роста оборота `yoy_turnover`. 

In [49]:
query = '''
WITH trade_region AS (
    SELECT YEAR(date_trade) AS year_trade,
           region_id,
           ROUND(SUM(turnover), 0) AS total_turnover,
           ROUND(AVG(avg_check), 0) AS avg_check
      FROM trade
     GROUP BY region_id, YEAR(date_trade)
),
region_details AS (
    SELECT id,
           region_trade
       FROM region_data_book
),
turnover_details AS (
    SELECT YEAR(date_trade) AS year_trade,
           ROUND(SUM(turnover), 0) AS year_total_turnover -- считаем годовой оборот
      FROM trade
     GROUP BY YEAR(date_trade)
),
trade_region_details AS (
SELECT trade_region.year_trade,
       region_trade,
       total_turnover,
       -- считаем долю в обороте
       ROUND(total_turnover*100 / year_total_turnover, 1) AS turnover_share,       
       -- считаем изменение доли в обороте с помощью оконной функции смещения
       ROUND((total_turnover/year_total_turnover - LAG(total_turnover/year_total_turnover, 1) OVER (
                                                       PARTITION BY region_trade  -- определение партиций по региону
                                                       ORDER BY year_trade))*100, 1
       ) AS yoy_turn_share,      
       -- считаем прирост оборота с помощью оконной функции смещения
       ROUND((total_turnover - LAG(total_turnover, 1) OVER (
                                   PARTITION BY region_trade -- определение партиций по региону
                                   ORDER BY year_trade)) / LAG(total_turnover, 1) OVER (
                                                               PARTITION BY region_trade -- определение партиций по региону
                                                               ORDER BY year_trade)*100, 0
       ) AS yoy_turnover,             
       avg_check,        
       -- считаем прирост среднего чека с помощью оконной функции смещения
       ROUND((avg_check - LAG(avg_check, 1) OVER (
                              PARTITION BY region_trade  -- определение партиций по региону
                              ORDER BY year_trade)) / LAG(avg_check, 1) OVER (
                                                          PARTITION BY region_trade  -- определение партиций по региону
                                                          ORDER BY year_trade)*100, 0
       ) AS yoy_avg_check     
  FROM trade_region
       INNER JOIN region_details AS r1
       ON trade_region.region_id = r1.id
       
       LEFT JOIN turnover_details AS t1
       ON trade_region.year_trade = t1.year_trade
),
trade_region_rnk AS (
SELECT *,
       -- считаем ранг с помощью ранжирующей оконной функции 
       RANK() OVER(ORDER BY yoy_turnover DESC) AS rnk 
  FROM trade_region_details
 WHERE trade_region_details.year_trade = 2022
)
SELECT *
  FROM trade_region_rnk
 WHERE rnk <= 5; -- вывод ТОП-5 регионов с наибольшими темпами роста оборота
'''
getQuery(query)

year_trade          region_trade  total_turnover  turnover_share  \
0        2022       Республика Крым          6855.0             0.1   
1        2022   Воронежская область         94726.0             2.0   
2        2022      Липецкая область         31533.0             0.7   
3        2022  Владимирская область         43751.0             0.9   
4        2022     Красноярский край         87188.0             1.8   

   yoy_turn_share  yoy_turnover  avg_check  yoy_avg_check  rnk  
0             0.1         123.0     3717.0           35.0    1  
1             0.5          80.0     3341.0           26.0    2  
2             0.1          63.0     3358.0           27.0    3  
3             0.2          59.0     3614.0           29.0    4  
4             0.3          58.0     3850.0           28.0    5

- **ТОП-5 регионов с наибольшими темпами роста оборота 2022 VS 2021:**
  - Республика Крым (прирост оборота 123%);
  - Воронежская область (прирост оборота 80%);
  - Липецкая область (прирост оборота 63%);
  - Владимирская область (прирост оборота 59%);
  - Красноярский край	 (прирост оборота 58%).

* Найдем 5 регионов с наименьшими темпами роста оборота `yoy_turnover`. 

In [50]:
query = '''
WITH trade_region AS (
    SELECT YEAR(date_trade) AS year_trade,
           region_id,
           ROUND(SUM(turnover), 0) AS total_turnover,
           ROUND(AVG(avg_check), 0) AS avg_check
      FROM trade
     GROUP BY region_id, YEAR(date_trade)
),
region_details AS (
    SELECT id,
           region_trade
       FROM region_data_book
),
turnover_details AS (
    SELECT YEAR(date_trade) AS year_trade,
           ROUND(SUM(turnover), 0) AS year_total_turnover -- считаем годовой оборот
      FROM trade
     GROUP BY YEAR(date_trade)
),
trade_region_details AS (
SELECT trade_region.year_trade,
       region_trade,
       total_turnover,
       -- считаем долю в обороте
       ROUND(total_turnover*100 / year_total_turnover, 1) AS turnover_share,       
       -- считаем изменение доли в обороте с помощью оконной функции смещения
       ROUND((total_turnover/year_total_turnover - LAG(total_turnover/year_total_turnover, 1) OVER (
                                                       PARTITION BY region_trade  -- определение партиций по региону
                                                       ORDER BY year_trade))*100, 1
       ) AS yoy_turn_share,      
       -- считаем прирост оборота с помощью оконной функции смещения
       ROUND((total_turnover - LAG(total_turnover, 1) OVER (
                                   PARTITION BY region_trade -- определение партиций по региону
                                   ORDER BY year_trade)) / LAG(total_turnover, 1) OVER (
                                                               PARTITION BY region_trade -- определение партиций по региону
                                                               ORDER BY year_trade)*100, 0
       ) AS yoy_turnover,             
       avg_check,        
       -- считаем прирост среднего чека с помощью оконной функции смещения
       ROUND((avg_check - LAG(avg_check, 1) OVER (
                              PARTITION BY region_trade  -- определение партиций по региону
                              ORDER BY year_trade)) / LAG(avg_check, 1) OVER (
                                                          PARTITION BY region_trade  -- определение партиций по региону
                                                          ORDER BY year_trade)*100, 0
       ) AS yoy_avg_check     
  FROM trade_region
       INNER JOIN region_details AS r1
       ON trade_region.region_id = r1.id
       
       LEFT JOIN turnover_details AS t1
       ON trade_region.year_trade = t1.year_trade
),
trade_region_rnk AS (
SELECT *,
       -- считаем ранг с помощью ранжирующей оконной функции 
       RANK() OVER(ORDER BY yoy_turnover ASC) AS rnk 
  FROM trade_region_details
 WHERE trade_region_details.year_trade = 2022
)
SELECT *
  FROM trade_region_rnk
 WHERE rnk <= 5; -- вывод ТОП-5 регионов с наименьшими темпами роста оборота
'''
getQuery(query)

year_trade                     region_trade  total_turnover  \
0        2022                 Хабаровский край         33954.0   
1        2022        Ненецкий автономный округ           904.0   
2        2022         Республика Саха (Якутия)         19113.0   
3        2022  Ямало-Ненецкий автономный округ         20974.0   
4        2022          Калининградская область         24494.0   

   turnover_share  yoy_turn_share  yoy_turnover  avg_check  yoy_avg_check  rnk  
0             0.7            -0.5         -22.0     4087.0            3.0    1  
1             0.0             0.0          -5.0     2329.0           -2.0    2  
2             0.4            -0.1          11.0     4352.0           25.0    3  
3             0.4            -0.1          15.0     4403.0           15.0    4  
4             0.5            -0.1          15.0     3782.0           22.0    4

Регионы с наименьшими темпами роста оборота 2022 VS 2021:
  - Хабаровский край (снижение оборота на 22%);
  - Ненецкий автономный округ (снижение оборота на 5%);


Остальные регионы показали положительный прирост оборота.

#### Вывод

**ТОП-10 регионов по доле в обороте:**
- ТОП-4 региона по доле в обороте в 2022 г. совпадают с 2021 г:
  - Москва - доля 19,6% (уменьшилась на 1,2 п.п. за год);
  - Московская область - доля 10,8% (не изменилась за год);
  - Санкт-Петербург - доля 7% (уменьшилась на 0,4 п.п. за год);
  - Краснодарский край - доля 4,7% (увеличилась на 0,3 п.п. за год);


- Ростовская область (доля увеличилась на 0,2 п.п. за год) в 2022 г. догнала Свердловскую область (доля не изменилась) - доля 2,6 п.п., 5-е и 6-е место.

- Нижегородская область - доля 2,1% (увеличилась на 0,1 п.п. за год) - 7-е место;
- Самарская область - доля 2% (не изменилась за год) - 8-е место;
- В 2022 г. **Воронежская область** показала **самый высокий прирост доли в обороте (0,5 п.п.)** и заняла 9-е место. В 2021 г. Воронежская область занимала 14-е место. 
- Республика Татарстан в 2022 г. опустилась на 10-е место. В 2021 г. в ТОП-10 входила республика Башкортостан.


- **Регионы наибольшими темпами роста оборота 2022 VS 2021**:
  - Республика Крым (прирост оборота 123%, оборот за 2022 - 6 855 млн руб.);
  - Воронежская область (прирост оборота 80%, оборот за 2022 - 94 726 млн руб.);
  - Липецкая область (прирост оборота 63%, оборот за 2022 - 31 533 млн руб.);
  - Владимирская область (прирост оборота 59%, оборот за 2022 - 43 751 млн руб.);
  - Красноярский край	 (прирост оборота 58%, оборот за 2022 - 87 188 млн руб.).


- **Регионы с наименьшими темпами роста оборота 2022 VS 2021**:
  - Хабаровский край (снижение оборота на 22%, оборот за 2022 - 33 954 млн руб.);
  - Ненецкий автономный округ (снижение оборота на 5%, оборот за 2022 - 904 млн руб.);
- Остальные регионы показали положительный прирост оборота.

### Определение ТОП-5 категорий товаров по обороту

* Найдем оборот (`total_turnover`), долю в обороте (`turnover_share`), изменение доли в обороте (`yoy_turn_share`), средний чек (`avg_check`), прирост оборота год к году (`yoy_turnover`), прирост среднего чека год к году (`yoy_avg_check`) в 2022 и 2021 годах и вывод топ-5 категорий по обороту.

**2021:**

In [52]:
query = '''
WITH trade_category AS (
    SELECT YEAR(date_trade) AS year_trade,
           category_id,
           ROUND(SUM(turnover), 0) AS total_turnover,
           ROUND(AVG(avg_check), 0) AS avg_check
      FROM trade
     GROUP BY category_id, YEAR(date_trade)
),
turnover_details AS (
    SELECT YEAR(date_trade) AS year_trade,
           ROUND(SUM(turnover), 0) AS year_total_turnover -- считаем годовой оборот
      FROM trade
     GROUP BY YEAR(date_trade)
),
category_details AS (
    SELECT category_id,
           category
       FROM category_base
),
trade_category_details AS (
SELECT trade_category.year_trade,
       category,
       total_turnover,
       -- считаем долю в обороте
       ROUND(total_turnover*100 / year_total_turnover, 1) AS turnover_share,       
       -- считаем изменение доли в обороте с помощью оконной функции смещения
       ROUND((total_turnover/year_total_turnover - LAG(total_turnover/year_total_turnover, 1) OVER (
                                                       PARTITION BY category  -- определение партиций категории 
                                                       ORDER BY year_trade))*100, 1
       ) AS yoy_turn_share,      
       -- считаем прирост оборота с помощью оконной функции смещения
       ROUND((total_turnover - LAG(total_turnover, 1) OVER (
                                   PARTITION BY category -- определение партиций по категории товаров
                                   ORDER BY year_trade)) / LAG(total_turnover, 1) OVER (
                                                               PARTITION BY category -- определение партиций по категории
                                                               ORDER BY year_trade)*100, 0
       ) AS yoy_turnover,             
       avg_check,        
       -- считаем прирост среднего чека с помощью оконной функции смещения
       ROUND((avg_check - LAG(avg_check, 1) OVER (
                              PARTITION BY category  -- определение партиций по категории товаров
                              ORDER BY year_trade)) / LAG(avg_check, 1) OVER (
                                                          PARTITION BY category  -- определение партиций по категории 
                                                          ORDER BY year_trade)*100, 0
       ) AS yoy_avg_check     
  FROM trade_category
       INNER JOIN category_details AS c1
       ON trade_category.category_id = c1.category_id
       
       LEFT JOIN turnover_details AS t1
       ON trade_category.year_trade = t1.year_trade
),
trade_category_rnk AS (
SELECT *,
       -- считаем ранг с помощью ранжирующей оконной функции 
       RANK() OVER(ORDER BY total_turnover DESC) AS rnk 
  FROM trade_category_details
 WHERE trade_category_details.year_trade = 2021
)
SELECT *
  FROM trade_category_rnk
 -- вывод ТОП-5 категорий товаров по обороту
 WHERE rnk <= 5; 
'''
getQuery(query).fillna('-')

year_trade                       category  total_turnover  turnover_share  \
0        2021     Цифровая и бытовая техника        851859.0            23.6   
1        2021       Мебель и товары для дома        613660.0            17.0   
2        2021                 Одежда и обувь        502391.0            13.9   
3        2021               Продукты питания        450103.0            12.5   
4        2021  Автозапчасти и автоаксессуары        269397.0             7.5   

  yoy_turn_share yoy_turnover  avg_check yoy_avg_check  rnk  
0              -            -    11482.0             -    1  
1              -            -     2363.0             -    2  
2              -            -     1668.0             -    3  
3              -            -     1103.0             -    4  
4              -            -     3609.0             -    5

**ТОП-5 категорий товаров по доле в обороте:**
  - Цифровая и бытовая техника - доля 23,6%;
  - Мебель и товары для дома - доля 17%;
  - Одежда и обувь	 - доля 13,9%;
  - Продукты питания - доля 12,5%;
  - Автозапчасти и автоаксессуары - доля 7,5%.

**2022:**

In [53]:
query = '''
WITH trade_category AS (
    SELECT YEAR(date_trade) AS year_trade,
           category_id,
           ROUND(SUM(turnover), 0) AS total_turnover,
           ROUND(AVG(avg_check), 0) AS avg_check
      FROM trade
     GROUP BY category_id, YEAR(date_trade)
),
turnover_details AS (
    SELECT YEAR(date_trade) AS year_trade,
           ROUND(SUM(turnover), 0) AS year_total_turnover -- считаем годовой оборот
      FROM trade
     GROUP BY YEAR(date_trade)
),
category_details AS (
    SELECT category_id,
           category
       FROM category_base
),
trade_category_details AS (
SELECT trade_category.year_trade,
       category,
       total_turnover,
       -- считаем долю в обороте
       ROUND(total_turnover*100 / year_total_turnover, 1) AS turnover_share,       
       -- считаем изменение доли в обороте с помощью оконной функции смещения
       ROUND((total_turnover/year_total_turnover - LAG(total_turnover/year_total_turnover, 1) OVER (
                                                       PARTITION BY category  -- определение партиций категории 
                                                       ORDER BY year_trade))*100, 1
       ) AS yoy_turn_share,      
       -- считаем прирост оборота с помощью оконной функции смещения
       ROUND((total_turnover - LAG(total_turnover, 1) OVER (
                                   PARTITION BY category -- определение партиций по категории товаров
                                   ORDER BY year_trade)) / LAG(total_turnover, 1) OVER (
                                                               PARTITION BY category -- определение партиций по категории
                                                               ORDER BY year_trade)*100, 0
       ) AS yoy_turnover,             
       avg_check,        
       -- считаем прирост среднего чека с помощью оконной функции смещения
       ROUND((avg_check - LAG(avg_check, 1) OVER (
                              PARTITION BY category  -- определение партиций по категории товаров
                              ORDER BY year_trade)) / LAG(avg_check, 1) OVER (
                                                          PARTITION BY category  -- определение партиций по категории 
                                                          ORDER BY year_trade)*100, 0
       ) AS yoy_avg_check     
  FROM trade_category
       INNER JOIN category_details AS c1
       ON trade_category.category_id = c1.category_id
       
       LEFT JOIN turnover_details AS t1
       ON trade_category.year_trade = t1.year_trade
),
trade_category_rnk AS (
SELECT *,
       -- считаем ранг с помощью ранжирующей оконной функции 
       RANK() OVER(ORDER BY total_turnover DESC) AS rnk 
  FROM trade_category_details
 WHERE trade_category_details.year_trade = 2022
)
SELECT *
  FROM trade_category_rnk
 -- вывод ТОП-5 категорий товаров по обороту
 WHERE rnk <= 5; 
'''
getQuery(query)

year_trade                    category  total_turnover  turnover_share  \
0        2022  Цифровая и бытовая техника       1021373.0            21.2   
1        2022    Мебель и товары для дома        865775.0            18.0   
2        2022              Одежда и обувь        710553.0            14.8   
3        2022            Продукты питания        653153.0            13.6   
4        2022          Красота и здоровье        375393.0             7.8   

   yoy_turn_share  yoy_turnover  avg_check  yoy_avg_check  rnk  
0            -2.4          20.0    12519.0            9.0    1  
1             1.0          41.0     1756.0          -26.0    2  
2             0.8          41.0     3361.0          101.0    3  
3             1.1          45.0     1110.0            1.0    4  
4             0.8          48.0     3951.0            1.0    5

**ТОП-5 категорий товаров по доле в обороте:**
  - Цифровая и бытовая техника - доля 21,2%;
  - Мебель и товары для дома - доля 18%;
  - Одежда и обувь	 - доля 14,8%;
  - Продукты питания - доля 13,6%;
  - Красота и здоровье - доля 7,8%.

В 2021 в ТОП-5 «Автозапчасти и автоаксессуары» с долей 7,5% вместо «Красота и здоровье», которая вошла в ТОП-5 в 2022 с долей 7,8%.

* Найдем 5 категорий товаров с наибольшими темпами роста оборота `yoy_turnover`. 

In [54]:
query = '''
WITH trade_category AS (
    SELECT YEAR(date_trade) AS year_trade,
           category_id,
           ROUND(SUM(turnover), 0) AS total_turnover,
           ROUND(AVG(avg_check), 0) AS avg_check
      FROM trade
     GROUP BY category_id, YEAR(date_trade)
),
turnover_details AS (
    SELECT YEAR(date_trade) AS year_trade,
           ROUND(SUM(turnover), 0) AS year_total_turnover -- считаем годовой оборот
      FROM trade
     GROUP BY YEAR(date_trade)
),
category_details AS (
    SELECT category_id,
           category
       FROM category_base
),
trade_category_details AS (
SELECT trade_category.year_trade,
       category,
       total_turnover,
       -- считаем долю в обороте
       ROUND(total_turnover*100 / year_total_turnover, 1) AS turnover_share,       
       -- считаем изменение доли в обороте с помощью оконной функции смещения
       ROUND((total_turnover/year_total_turnover - LAG(total_turnover/year_total_turnover, 1) OVER (
                                                       PARTITION BY category  -- определение партиций категории 
                                                       ORDER BY year_trade))*100, 1
       ) AS yoy_turn_share,      
       -- считаем прирост оборота с помощью оконной функции смещения
       ROUND((total_turnover - LAG(total_turnover, 1) OVER (
                                   PARTITION BY category -- определение партиций по категории товаров
                                   ORDER BY year_trade)) / LAG(total_turnover, 1) OVER (
                                                               PARTITION BY category -- определение партиций по категории
                                                               ORDER BY year_trade)*100, 0
       ) AS yoy_turnover,             
       avg_check,        
       -- считаем прирост среднего чека с помощью оконной функции смещения
       ROUND((avg_check - LAG(avg_check, 1) OVER (
                              PARTITION BY category  -- определение партиций по категории товаров
                              ORDER BY year_trade)) / LAG(avg_check, 1) OVER (
                                                          PARTITION BY category  -- определение партиций по категории 
                                                          ORDER BY year_trade)*100, 0
       ) AS yoy_avg_check     
  FROM trade_category
       INNER JOIN category_details AS c1
       ON trade_category.category_id = c1.category_id
       
       LEFT JOIN turnover_details AS t1
       ON trade_category.year_trade = t1.year_trade
),
trade_category_rnk AS (
SELECT *,
       -- считаем ранг с помощью ранжирующей оконной функции 
       RANK() OVER(ORDER BY yoy_turnover DESC) AS rnk 
  FROM trade_category_details
 WHERE trade_category_details.year_trade = 2022
)
SELECT *
  FROM trade_category_rnk
 -- вывод ТОП-5 категорий товаров с наибольшими темпами роста оборота
 WHERE rnk <= 5; 
'''
getQuery(query)

year_trade                       category  total_turnover  turnover_share  \
0        2022       Ювелирные изделия и часы         36850.0             0.8   
1        2022  Инструменты и садовая техника        118509.0             2.5   
2        2022                     Аксессуары        108244.0             2.3   
3        2022                    Спорттовары        127215.0             2.6   
4        2022                    Зоомагазины         76096.0             1.6   

   yoy_turn_share  yoy_turnover  avg_check  yoy_avg_check  rnk  
0             0.4         205.0     5923.0           -4.0    1  
1             1.3         191.0     1967.0          -15.0    2  
2             0.7          88.0     5208.0          197.0    3  
3             0.7          85.0     5318.0           84.0    4  
4             0.3          65.0     2646.0           45.0    5

- **ТОП-5 категорий товаров с наибольшими темпами роста оборота 2022 VS 2021:**
  - Ювелирные изделия и часы (прирост оборота 205%);
  - Инструменты и садовая техника (прирост оборота 191%);
  - Аксессуары (прирост оборота 88%);
  - Спорттовары (прирост оборота 85%);
  - Зоомагазины	(прирост оборота 65%).

* Найдем 5 категорий товаров с наименьшими темпами роста оборота `yoy_turnover`. 

In [55]:
query = '''
WITH trade_category AS (
    SELECT YEAR(date_trade) AS year_trade,
           category_id,
           ROUND(SUM(turnover), 0) AS total_turnover,
           ROUND(AVG(avg_check), 0) AS avg_check
      FROM trade
     GROUP BY category_id, YEAR(date_trade)
),
turnover_details AS (
    SELECT YEAR(date_trade) AS year_trade,
           ROUND(SUM(turnover), 0) AS year_total_turnover -- считаем годовой оборот
      FROM trade
     GROUP BY YEAR(date_trade)
),
category_details AS (
    SELECT category_id,
           category
       FROM category_base
),
trade_category_details AS (
SELECT trade_category.year_trade,
       category,
       total_turnover,
       -- считаем долю в обороте
       ROUND(total_turnover*100 / year_total_turnover, 1) AS turnover_share,       
       -- считаем изменение доли в обороте с помощью оконной функции смещения
       ROUND((total_turnover/year_total_turnover - LAG(total_turnover/year_total_turnover, 1) OVER (
                                                       PARTITION BY category  -- определение партиций категории 
                                                       ORDER BY year_trade))*100, 1
       ) AS yoy_turn_share,      
       -- считаем прирост оборота с помощью оконной функции смещения
       ROUND((total_turnover - LAG(total_turnover, 1) OVER (
                                   PARTITION BY category -- определение партиций по категории товаров
                                   ORDER BY year_trade)) / LAG(total_turnover, 1) OVER (
                                                               PARTITION BY category -- определение партиций по категории
                                                               ORDER BY year_trade)*100, 0
       ) AS yoy_turnover,             
       avg_check,        
       -- считаем прирост среднего чека с помощью оконной функции смещения
       ROUND((avg_check - LAG(avg_check, 1) OVER (
                              PARTITION BY category  -- определение партиций по категории товаров
                              ORDER BY year_trade)) / LAG(avg_check, 1) OVER (
                                                          PARTITION BY category  -- определение партиций по категории 
                                                          ORDER BY year_trade)*100, 0
       ) AS yoy_avg_check     
  FROM trade_category
       INNER JOIN category_details AS c1
       ON trade_category.category_id = c1.category_id
       
       LEFT JOIN turnover_details AS t1
       ON trade_category.year_trade = t1.year_trade
),
trade_category_rnk AS (
SELECT *,
       -- считаем ранг с помощью ранжирующей оконной функции 
       RANK() OVER(ORDER BY yoy_turnover ASC) AS rnk 
  FROM trade_category_details
 WHERE trade_category_details.year_trade = 2022
)
SELECT *
  FROM trade_category_rnk
 -- вывод ТОП-5 категорий товаров с наименьшими темпами роста оборота
 WHERE rnk <= 5; 
'''
getQuery(query)

year_trade                       category  total_turnover  turnover_share  \
0        2022                  Прочие товары         50442.0             1.0   
1        2022  Автозапчасти и автоаксессуары        254923.0             5.3   
2        2022                          Книги         60537.0             1.3   
3        2022     Цифровая и бытовая техника       1021373.0            21.2   
4        2022                Подарки и цветы         36808.0             0.8   

   yoy_turn_share  yoy_turnover  avg_check  yoy_avg_check  rnk  
0            -2.8         -63.0     1469.0           27.0    1  
1            -2.2          -5.0     4251.0           18.0    2  
2            -0.2          17.0      680.0           -3.0    3  
3            -2.4          20.0    12519.0            9.0    4  
4             0.0          33.0     2846.0           27.0    5

- **Категории товаров с наименьшими темпами роста оборота 2022 VS 2021:**
  - Прочие товары (снижение оборота на 63%);
  - Автозапчасти и автоаксессуары (снижение оборота на 5%);
  - Книги (прирост оборота 17%);
  - Цифровая и бытовая техника (прирост оборота 20%);
  - Подарки и цветы	(прирост оборота 33%).

#### Вывод

**ТОП-5 категорий товаров по доле в обороте:**
- ТОП-4 категории по доле в обороте в 2022 г. совпадают с 2021 г:
  - Цифровая и бытовая техника - доля 21,2% (уменьшилась на 2,4 п.п. за год);
  - Мебель и товары для дома - доля 18% (увеличилась на 1 п.п. за год);
  - Одежда и обувь	 - доля 14,8% (увеличилась на 0,8 п.п. за год);
  - Продукты питания - доля 13,6% (увеличилась на 1,1 п.п. за год);
  
- В 2021 в ТОП-5 «Автозапчасти и автоаксессуары» с долей 7,5% вместо «Красота и здоровье», которая вошла в ТОП-5 в 2022 с долей 7,8%.


- **ТОП-5 категорий товаров с наибольшими темпами роста оборота 2022 VS 2021**:
  - Ювелирные изделия и часы (прирост оборота 205%, оборот за 2022 - 36 850 млн руб.);
  - Инструменты и садовая техника (прирост оборота 191%, оборот за 2022 - 118 509 млн руб.);
  - Аксессуары (прирост оборота 88%, оборот за 2022 - 108 244 млн руб.);
  - Спорттовары (прирост оборота 85%, оборот за 2022 - 127 215 млн руб.);
  - Зоомагазины	(прирост оборота 65%, оборот за 2022 - 76 096 млн руб.).
  
  
- **Категории товаров с наименьшими темпами роста оборота 2022 VS 2021**:
  - Прочие товары (снижение оборота на 63%, оборот за 2022 - 50 442 млн руб.);
  - Автозапчасти и автоаксессуары (снижение оборота на 5%, оборот за 2022 - 25 4923 млн руб.);
- Остальные категории показали положительный прирост оборота.

### Сравнение среднего чека выбранного региона со средним чеком группы регионов по численности населения

* Определение группы по населению в зависимости от численности населения в поле `population_group` в таблице `population_base`:
  - "[1]<100 тыс.чел.";
  - "[2] 100-500 тыс.чел.";
  - "[3] 500-1 млн чел.";
  - "[4]>1 млн чел.".

Создадим поле `population_group`.

In [ ]:
query = '''
ALTER TABLE population_base
ADD COLUMN population_group VARCHAR(40);
'''

Выполнение запроса.

In [ ]:
with cnx.connect() as connection:
    connection.execute(query)

Определим группы по населению в поле `population_group`.

In [ ]:
query = '''
UPDATE population_base
SET population_group = IF(population <= 100000, "[1]<100 тыс.чел.",
                       IF(population > 100000 AND population <= 500000, "[2] 100-500 тыс.чел.",
                       IF(population > 1000000, "[4]>1 млн чел.", "[3] 500-1 млн чел."))
                          );
''' 

Выполнение запроса.

In [ ]:
with cnx.connect() as connection:
    connection.execute(query)

Выведем первые 5 строк таблицы `population_base`.

In [56]:
query = '''
SELECT *
  FROM population_base
 LIMIT 5; -- вывод первых 5 строк
'''
getQuery(query)

region  population    population_group
0             Алтайский край     2268179      [4]>1 млн чел.
1           Амурская область      772525  [3] 500-1 млн чел.
2  Архангельская обл.без НАО     1069782      [4]>1 млн чел.
3       Астраханская область      989430  [3] 500-1 млн чел.
4       Белгородская область     1531917      [4]>1 млн чел.

* Найдем оборот (`total_turnover`), средний чек (`avg_check`), прирост оборота год к году (`yoy_turnover`), прирост среднего чека год к году (`yoy_avg_check`) в 2022 и 2021 годах по группам по населению.

In [60]:
query = '''
WITH trade_region AS (
    SELECT YEAR(date_trade) AS year_trade,
           region_id,
           SUM(turnover) AS total_turnover,
           AVG(avg_check) AS avg_check
      FROM trade
     GROUP BY region_id, YEAR(date_trade)
),
region_details AS (
    SELECT id,
           region_population
      FROM region_data_book
),
population_details AS (
    SELECT population_group,
           region
      FROM population_base
),
trade_region_details AS (
    SELECT year_trade,
           population_group,
           ROUND(SUM(total_turnover), 0) AS total_turnover,
           ROUND(AVG(avg_check), 0) AS avg_check
      FROM trade_region
           INNER JOIN region_details AS reg
           ON trade_region.region_id = reg.id
       
           INNER JOIN population_details AS p
           ON reg.region_population = p.region
     GROUP BY population_group, year_trade
)
SELECT year_trade,
       population_group,
       total_turnover,       
       -- считаем прирост оборота с помощью оконной функции смещения
       ROUND((total_turnover - LAG(total_turnover, 1) OVER (
                                   PARTITION BY population_group -- определение партиций по группе населения
                                   ORDER BY year_trade)) / LAG(total_turnover, 1) OVER (
                                                               -- определение партиций по группе населения
                                                               PARTITION BY population_group 
                                                               ORDER BY year_trade)*100, 0
       ) AS yoy_turnover,
       avg_check,       
       -- считаем прирост среднего чека с помощью оконной функции смещения
       ROUND((avg_check - LAG(avg_check, 1) OVER (
                              PARTITION BY population_group -- определение партиций по группе населения
                              ORDER BY year_trade)) / LAG(avg_check, 1) OVER (
                                                          -- определение партиций по группе населения
                                                          PARTITION BY population_group 
                                                          ORDER BY year_trade)*100, 0
        ) AS yoy_avg_check
  FROM trade_region_details
 ORDER BY year_trade DESC;
'''
getQuery(query).fillna('-')

year_trade      population_group  total_turnover yoy_turnover  avg_check  \
0        2022      [1]<100 тыс.чел.          3949.0         15.0     3239.0   
1        2022  [2] 100-500 тыс.чел.         64705.0         28.0     3079.0   
2        2022    [3] 500-1 млн чел.        389151.0         31.0     3399.0   
3        2022        [4]>1 млн чел.       4351257.0         34.0     3542.0   
4        2021      [1]<100 тыс.чел.          3428.0            -     3082.0   
5        2021  [2] 100-500 тыс.чел.         50704.0            -     2666.0   
6        2021    [3] 500-1 млн чел.        296170.0            -     2724.0   
7        2021        [4]>1 млн чел.       3253999.0            -     2960.0   

  yoy_avg_check  
0           5.0  
1          15.0  
2          25.0  
3          20.0  
4             -  
5             -  
6             -  
7             -

- Прирост оборота и среднего чека по регионам группы с численностью населения более 1 млн чел. 2022 VS 2021 совпадает с общими показателями по всем регионам.
- Прирост оборота группы с численностью населения от 500 тыс. до 1 млн чел. 2022 VS 2021 на 2% ниже общего, а прирост среднего чека на 5% выше.
- Прирост оборота и среднего чека группы с численностью населения от 100 до 500 тыс.чел. 2022 VS 2021 ниже общих показателей на 5%.
- В группе с численностью населения до 100 тыс.чел. прирост оборота 2022 VS 2021 почти в два раза ниже общего показателя, а прирост среднего чека ниже общего показателя в 4 раза.

* Определение среднего чека `avg_check_group` по группе по населению, отклонения среднего чека региона от среднего чека по по группе по населению `deviation_avg_check`.

**2021:**

In [69]:
query = '''
WITH trade_region AS (
    SELECT YEAR(date_trade) AS year_trade,
           region_id,
           SUM(turnover) AS total_turnover,
           AVG(avg_check) AS avg_check
      FROM trade
     GROUP BY region_id, YEAR(date_trade)
),
region_details AS (
    SELECT id,
           region_population,
           region_trade
      FROM region_data_book
),
population_details AS (
    SELECT population_group,
           region
      FROM population_base
),
trade_region_details AS (
    SELECT year_trade,
           region_trade,
           population_group,
           ROUND(SUM(total_turnover), 0) AS total_turnover,
           ROUND(AVG(avg_check), 0) AS avg_check
      FROM trade_region
           INNER JOIN region_details AS reg
           ON trade_region.region_id = reg.id
       
           INNER JOIN population_details AS p
           ON reg.region_population = p.region
     WHERE year_trade = 2021
     GROUP BY population_group, region_trade, year_trade
)
SELECT year_trade,
       region_trade,
       population_group,
       avg_check,
       -- считаем средний чек по группе населения с помощью агрегирующей оконной функции
       ROUND(AVG(avg_check) OVER (
                 PARTITION BY population_group -- определение партиций по группе населения
                 ORDER BY year_trade), 0
       ) AS avg_check_group,
       -- считаем отклонение среднего чека региона от среднего чек по группе с помощью агрегирующей оконной функции
       ROUND((avg_check - AVG(avg_check) OVER (
                              PARTITION BY population_group -- определение партиций по группе населения
                              ORDER BY year_trade)) / AVG(avg_check) OVER (
                                                          PARTITION BY population_group -- определение партиций по группе
                                                          ORDER BY year_trade)*100, 0
       ) AS "deviation_avg_check"
  FROM trade_region_details
 ORDER BY deviation_avg_check DESC;
'''
getQuery(query)

year_trade                     region_trade      population_group  \
0         2021              Сахалинская область  [2] 100-500 тыс.чел.   
1         2021              Республика Дагестан        [4]>1 млн чел.   
2         2021                           Москва        [4]>1 млн чел.   
3         2021                  Камчатский край  [2] 100-500 тыс.чел.   
4         2021  Ямало-Ненецкий автономный округ    [3] 500-1 млн чел.   
..         ...                              ...                   ...   
80        2021              Республика Мордовия    [3] 500-1 млн чел.   
81        2021        Ненецкий автономный округ      [1]<100 тыс.чел.   
82        2021              Республика Калмыкия  [2] 100-500 тыс.чел.   
83        2021              Республика Марий Эл    [3] 500-1 млн чел.   
84        2021                 Республика Алтай  [2] 100-500 тыс.чел.   

    avg_check  avg_check_group  deviation_avg_check  
0      4265.0           2666.0                 60.0  
1      4612.0           2960.0                 56.0  
2      4333.0           2960.0                 46.0  
3      3788.0           2666.0                 42.0  
4      3833.0           2724.0                 41.0  
..        ...              ...                  ...  
80     2158.0           2724.0                -21.0  
81     2385.0           3083.0                -23.0  
82     2030.0           2666.0                -24.0  
83     2063.0           2724.0                -24.0  
84     1975.0           2666.0                -26.0  

[85 rows x 6 columns]

- Регионами с самым большим положительным отклонением среднего чека от среднего чека по группе населения являются Сахалинская область (60%), группа 100-500 тыс.чел и Республика Дагестан (56%), группа >1 млн чел.
- Регионами с самым большим отрицательным отклонением среднего чека от среднего чека по группе населения являются Республика Алтай (-26%), группа 100-500 тыс.чел, Республика Марий Эл (-24%), группа 500-1 млн чел и Республика Калмыкия (-24%), группа 100-500 тыс.чел.

**2022:**

In [67]:
query = '''
WITH trade_region AS (
    SELECT YEAR(date_trade) AS year_trade,
           region_id,
           SUM(turnover) AS total_turnover,
           AVG(avg_check) AS avg_check
      FROM trade
     GROUP BY region_id, YEAR(date_trade)
),
region_details AS (
    SELECT id,
           region_population,
           region_trade
      FROM region_data_book
),
population_details AS (
    SELECT population_group,
           region
      FROM population_base
),
trade_region_details AS (
    SELECT year_trade,
           region_trade,
           population_group,
           ROUND(SUM(total_turnover), 0) AS total_turnover,
           ROUND(AVG(avg_check), 0) AS avg_check
      FROM trade_region
           INNER JOIN region_details AS reg
           ON trade_region.region_id = reg.id
       
           INNER JOIN population_details AS p
           ON reg.region_population = p.region
     WHERE year_trade = 2022
     GROUP BY population_group, region_trade, year_trade
)
SELECT year_trade,
       region_trade,
       population_group,
       avg_check,
       -- считаем средний чек по группе населения с помощью агрегирующей оконной функции
       ROUND(AVG(avg_check) OVER (
                 PARTITION BY population_group -- определение партиций по группе населения
                 ORDER BY year_trade), 0
       ) AS avg_check_group,
       -- считаем отклонение среднего чека региона от среднего чек по группе с помощью агрегирующей оконной функции
       ROUND((avg_check - AVG(avg_check) OVER (
                              PARTITION BY population_group -- определение партиций по группе населения
                              ORDER BY year_trade)) / AVG(avg_check) OVER (
                                                          PARTITION BY population_group -- определение партиций по группе
                                                          ORDER BY year_trade)*100, 0
       ) AS "deviation_avg_check"
  FROM trade_region_details
 ORDER BY deviation_avg_check DESC;
'''
getQuery(query)

year_trade                     region_trade      population_group  \
0         2022              Сахалинская область  [2] 100-500 тыс.чел.   
1         2022                  Камчатский край  [2] 100-500 тыс.чел.   
2         2022  Ямало-Ненецкий автономный округ    [3] 500-1 млн чел.   
3         2022       Чукотский автономный округ      [1]<100 тыс.чел.   
4         2022         Республика Саха (Якутия)    [3] 500-1 млн чел.   
..         ...                              ...                   ...   
80        2022                 Республика Алтай  [2] 100-500 тыс.чел.   
81        2022                  Республика Тыва  [2] 100-500 тыс.чел.   
82        2022     Еврейская автономная область  [2] 100-500 тыс.чел.   
83        2022        Ненецкий автономный округ      [1]<100 тыс.чел.   
84        2022              Республика Калмыкия  [2] 100-500 тыс.чел.   

    avg_check  avg_check_group  deviation_avg_check  
0      4920.0           3079.0                 60.0  
1      4880.0           3079.0                 58.0  
2      4403.0           3399.0                 30.0  
3      4150.0           3240.0                 28.0  
4      4352.0           3399.0                 28.0  
..        ...              ...                  ...  
80     2304.0           3079.0                -25.0  
81     2266.0           3079.0                -26.0  
82     2227.0           3079.0                -28.0  
83     2329.0           3240.0                -28.0  
84     2171.0           3079.0                -29.0  

[85 rows x 6 columns]

- Регионами с самым большим положительным отклонением среднего чека от среднего чека по группе населения являются Сахалинская область (60%) и Камчатский край (58%), группа 100-500 тыс.чел.
- Регионами с самым большим отрицательным отклонением среднего чека от среднего чека по группе населения являются Республика Калмыкия (-29%), группа 100-500 тыс.чел, Ненецкий автономный округ (-28%), группа <100 тыс.чел и Еврейская автономная область (-28%), группа 100-500 тыс.чел.

#### Вывод

- Прирост оборота и среднего чека по регионам группы с численностью населения более 1 млн чел. 2022 VS 2021 совпадает с общими показателями по всем регионам.
- В группах с численностью населения от 500 тыс. до 1 млн чел. и от 100 до 500 тыс.чел. прирост оборота и среднего чека 2022 VS 2021 отличается от общих показателей не более, чем на 5%.

- В группе с численностью населения до 100 тыс.чел. прирост оборота 2022 VS 2021 почти в два раза ниже общего показателя, а прирост среднего чека ниже общего показателя в 4 раза.

- Регионом с самым большим положительным отклонением среднего чека от среднего чека по группе населения в 2021 и 2022 является Сахалинская область (60%), группа 100-500 тыс.чел.
- Регионами с самым большим отрицательным отклонением среднего чека от среднего чека по группе населения в 2021 является Республика Алтай (-26%), группа 100-500 тыс.чел, в 2022 - Республика Калмыкия (-29%), группа 100-500 тыс.чел.

- В малонаселенных регионах меньше клиентов, плохо развита сеть пунктов выдачи, затруднена доставка, что сказывается на обороте и темпах роста.

### Создание представления

* Создадим представление `trade_v`.

In [ ]:
query = '''
CREATE VIEW trade_v AS
    SELECT
        t.id,
        t.region_id,
        r.region_trade,
        t.date_trade,
        t.turnover,
        t.avg_check,
        c.category,
        p.population_group
      FROM trade AS t
           LEFT JOIN region_data_book AS r                        
           ON t.region_id = r.id
       
           LEFT JOIN category_base AS c                        
           ON t.category_id = c.category_id
       
           LEFT JOIN population_base AS p                        
           ON r.region_population = p.region;
'''

Выполнение запроса.

In [ ]:
with cnx.connect() as connection:
    connection.execute(query)

* Напишем запрос на выгрузку первых 5 строк представления `trade_v`.

In [4]:
query = '''
SELECT * 
  FROM trade_v
 LIMIT 5
'''
getQuery(query)

id  region_id          region_trade  date_trade  turnover  avg_check  \
0   1          8  Белгородская область  2021-01-01    255.86       2162   
1   2          9      Брянская область  2021-01-01    277.04       2751   
2   3         83  Владимирская область  2021-01-01    354.88       2858   
3   4         86   Воронежская область  2021-01-01    608.76       1702   
4   5         22    Ивановская область  2021-01-01    209.59       2261   

                        category    population_group  
0  Автозапчасти и автоаксессуары      [4]>1 млн чел.  
1  Автозапчасти и автоаксессуары      [4]>1 млн чел.  
2  Автозапчасти и автоаксессуары      [4]>1 млн чел.  
3  Автозапчасти и автоаксессуары      [4]>1 млн чел.  
4  Автозапчасти и автоаксессуары  [3] 500-1 млн чел.

Представление для отчета создано.

## Общий вывод

- Для проведения исследования из открытых источников собраны данные об обороте и среднем чеке внутренней интернет-торговли по категориям товаров в регионах за 2021 - 2022, а также о численности населения регионов РФ на 01.01.2022.
- Подготовлен датасет и создана база данных на MySQL в клиенте DBeaver.
- Написаны необходимые SQL-запросы и создано представление, на основании которого разработан отчет в Power BI Desktop.


Выводы по исследованию:
- Оборот внутренней интернет-торговли в России за 2022 год увеличился на 33% по сравнению с показателем 2021 года и составил 4,81 трлн рублей.
- Средний чек увеличился на 20% по сравнению с 2021 годом и составил 3 447 руб.


- На потребительское поведение в 2022 году повлияли политические события.
  - Снижение оборота в 1-м и 2-м квартале может быть обусловлено тем, что: 
    - Сократились объемы производства, строительства, недвижимости.
    - Многие компании покинули российский рынок.
    - Кто-то в начале весны стал покупать меньше из-за неопределенности и возможности потерять работу, сложности строить планы на будущее. 
  - Летом покупательская активность в интернете постепенно начала восстанавливаться.
  - В 4м-квартале покупательская активность увеличилась - влияние сезонного повышения спроса в связи с распродажами, покупкой подарков к новогодним праздникам.
  
  
- **ТОП-5 регионов по доле в обороте в 2022**:
  - Москва - доля 19,6%;
  - Московская область - доля 10,8%;
  - Санкт-Петербург - доля 7%;
  - Краснодарский край - доля 4,7%;
  - Ростовская область - доля 2,6%;
- В 2022 г. Воронежская область показала самый высокий прирост доли в обороте (0,5 п.п.) и заняла 9-е место. В 2021 г. Воронежская область занимала 14-е место. 


- **Регионы наибольшими темпами роста оборота 2022 VS 2021**:
  - Республика Крым (прирост оборота 123%);
  - Воронежская область (прирост оборота 80%);
  - Липецкая область (прирост оборота 63%);
  
 
- **ТОП-5 категорий товаров по доле в обороте в 2022**:
  - Цифровая и бытовая техника - доля 21,2%;
  - Мебель и товары для дома - доля 18%;
  - Одежда и обувь	 - доля 14,8%;
  - Продукты питания - доля 13,6%;
  - «Красота и здоровье» - доля 7,8%% 
- В 2021 в ТОП-5 «Автозапчасти и автоаксессуары» с долей 7,5% вместо «Красота и здоровье».


- **ТОП-5 категорий товаров с наибольшими темпами роста оборота 2022 VS 2021**:
  - Ювелирные изделия и часы (прирост оборота 205%);
  - Инструменты и садовая техника (прирост оборота 191%);
  - Аксессуары (прирост оборота 88%);
  - Спорттовары (прирост оборота 85%);
  - Зоомагазины	(прирост оборота 65%).
  
  
- Регионом с самым большим положительным отклонением среднего чека от среднего чека по группе населения в 2021 и 2022 является Сахалинская область (60%), группа 100-500 тыс.чел.
- Регионом с самым большим отрицательным отклонением среднего чека от среднего чека по группе населения в 2021 является Республика Алтай (-26%), группа 100-500 тыс.чел, в 2022 - Республика Калмыкия (-29%), группа 100-500 тыс.чел.
- В малонаселенных регионах меньше клиентов, плохо развита сеть пунктов выдачи, затруднена доставка, что сказывается на обороте и темпах роста.